ca/ca.go

```diff
@@ -571,17 +573,44 @@ func (ca *CertificateAuthorityImpl) IssuePrecertificate(ctx context.Context, iss
 	}, nil
 }
 
-// IssueCertificateForPrecertificate takes a precertificate and a set of SCTs for that precertificate
-// and uses the signer to create and sign a certificate from them. The poison extension is removed
-// and a SCT list extension is inserted in its place. Except for this and the signature the certificate
-// exactly matches the precertificate. After the certificate is signed a OCSP response is generated
-// and the response and certificate are stored in the database.
+// IssueCertificateForPrecertificate takes a precertificate and a set
+// of SCTs for that precertificate and uses the signer to create and
+// sign a certificate from them. The poison extension is removed and a
+// SCT list extension is inserted in its place. Except for this and the
+// signature the certificate exactly matches the precertificate. After
+// the certificate is signed a OCSP response is generated and the
+// response and certificate are stored in the database.
+//
+// It's critical not to sign two different final certificates for the same
+// precertificate. This can happen, for instance, if the caller provides a
+// different set of SCTs on subsequent calls to  IssueCertificateForPrecertificate.
+// We rely on the RA not to call IssueCertificateForPrecertificate twice for the
+// same serial. This is accomplished by the fact that
+// IssueCertificateForPrecertificate is only ever called in a straight-through
+// RPC path without retries. If there is any error, including a networking
+// error, the whole certificate issuance attempt fails and any subsequent
+// issuance will use a different serial number.
+//
+// We also check that the provided serial number does not already exist as a
+// final certificate, but this is just a belt-and-suspenders measure, since
+// there could be race conditions where two goroutines are issuing for the same
+// serial number at the same time.
 func (ca *CertificateAuthorityImpl) IssueCertificateForPrecertificate(ctx context.Context, req *caPB.IssueCertificateForPrecertificateRequest) (core.Certificate, error) {
 	emptyCert := core.Certificate{}
 	precert, err := x509.ParseCertificate(req.DER)
 	if err != nil {
 		return emptyCert, err
 	}
+
+	serialHex := core.SerialToString(precert.SerialNumber)
+	_, err = ca.sa.GetCertificate(ctx, serialHex)
+	if err == nil {
+		err = berrors.InternalServerError("issuance of duplicate final certificate requested: %s", serialHex)
+		ca.log.AuditErr(err.Error())
+		return emptyCert, err
+	} else if !db.IsNoRows(err) {
+		return emptyCert, fmt.Errorf("error checking for duplicate issuance of %s: %s", serialHex, err)
+	}
```

Comment:
It looks like `sa.GetCertificate` uses `db.IsNoRows` internally to transform a no-rows `db.ErrDatabaseOp` into a `berrors.NotFoundError` that gets returned indicating the serial wasn't found:

https://github.com/letsencrypt/boulder/blob/e9e15c9a832282c068237a3cdb31c1d91e0de986/sa/sa.go#L361-L364

The `berrors.NotFoundError` instance will always get `false` from `db.IsNoRows` when your code checks it here and that ends up breaking CI when a no-results response from `ca.sa.GetCertificate` is treated as an unexpected error checking for the duplicate:
```
I012549 boulder-ra n83H5Aw [AUDIT] Certificate request - error JSON={"ID":"Z1UKmuLppZ5PLNB3FvbH3C2Ch2NCIaSP0xK12v38yR4","Requester":2,"OrderID":1,"VerifiedFields":["subject.commonName","subjectAltName"],"Names":["rand.3d2a8b8c.xyz"],"NotBefore":"0001-01-01T00:00:00Z","NotAfter":"0001-01-01T00:00:00Z","RequestTime":"2019-06-10T01:24:34Z","ResponseTime":"2019-06-10T01:24:34Z","Error":"issuing certificate for precertificate: rpc error: code = Unknown desc = error checking for duplicate issuance of ff6b088b57833e5ac2597d87f2365c8f85fb: certificate with serial \"ff6b088b57833e5ac2597d87f2365c8f85fb\" not found","Authorizations":{"rand.3d2a8b8c.xyz":{"ID":"1","ChallengeType":"dns-01"}}}
```

I think you want this instead:
```
} else if !berrors.Is(err, berrors.NotFound) {
...
```

Reply:
I tried this and got:
```
    ca_test.go:893: Failed to issue cert from precert: error checking for duplicate issuance of 11d8536fc4c4b8936a42e9cbb0c7102a1113: failed to : sql: no rows in result set
```

I think I might need to plumb through a little more. It looks like this is one of the new wrapped database errors. Will give this a shot tomorrow. Pushed my work-in-progress for now (will fail tests).